### PR TITLE
User authentication for GitHub via OAuth 2.0

### PR DESCRIPTION
**Pull request by kuries**
Related to: #11 
Parent: #28

Tasks:

- [x] Routes for GitHub OAuth in backend
- [x] Webpage to handle redirect from GitHub server
- [x] Context hooks to store github user id
- [x] Store github user id in server sessions
- [x] Store github user id and access tokens in account data (#31)

Reference for GitHub OAuth: https://docs.github.com/en/developers/apps/building-oauth-apps/authorizing-oauth-apps#web-application-flow

### CODE REVIEW

File: frontend/src/Pages/GithubOauth.js [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+            .catch( (err) => {
+                console.error(err);
+            });
+        }
+        const resp = sendCode();
+        resp.then( (resp) => {
+            navigate('/');
+        });
+    });
+    return (
```
Comment (Half-Shot) [inline frontend/src/Pages/GithubOauth.js:34]: Is this a valid state? Should we have an error code here or a message like "you are being redirected"?

File: frontend/src/Pages/GithubOauth.js [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+        async function sendCode(){
+            await axios.post('api/github/login', {
+                "code": loginCode
+            })
+            .catch( (err) => {
+                console.error(err);
+            });
+        }
+        const resp = sendCode();
+        resp.then( (resp) => {
```
Comment (Half-Shot) [inline frontend/src/Pages/GithubOauth.js:30]: You don't use the response here, so this is just overlapping your const.
```suggestion
        resp.then( () => {
```

File: frontend/src/Components/LayoutsWithNavbar.js [hunk 3]
```diff
@@ -5,16 +5,18 @@ import { GlobalContext } from '../GlobalContext';
...
       const fetchCurrentUser = async () => {
         const resp = await axios.get('api/current-user');
         console.log(resp.data);
         if(resp.data.matrix_user_id !== null){
-            const user_id = resp.data.matrix_user_id;
-            setMatrixUserId(user_id);
+            const matrix_user_id = resp.data.matrix_user_id;
+            const github_user_id = resp.data.github_user_id ? resp.data.github_user_id : "";
+            setMatrixUserId(matrix_user_id);
+            setGithubUserId(github_user_id);
```
Comment (Half-Shot) [inline frontend/src/Components/LayoutsWithNavbar.js:19]: mixture of casing, would prefer camelCase

```suggestion
            const matrixUseId = resp.data.matrix_user_id;
            const githubUserId = resp.data.github_user_id ? resp.data.github_user_id : "";
            setMatrixUserId(matrixUseId);
            setGithubUserId(githubUserId);
```

File: frontend/src/Pages/GithubOauth.js (hunk 1 above)
Comment (kuries) [reply frontend/src/Pages/GithubOauth.js:34]: I've added a text message in HTML for it 👍

https://github.com/MatrixDoctors/matrix-cerberus/blob/398e3af73cfb6f7867b9553e7d9de7d8d8c86300/frontend/src/Pages/GithubOauth.js#L32-L35

File: frontend/src/Pages/GithubOauth.js (hunk 2 above)
Comment (kuries) [reply frontend/src/Pages/GithubOauth.js:30]: Done!

File: frontend/src/Pages/GithubOauth.js [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+            navigate('/');
+        });
```
[the remark this answers is not in the document]
Comment (kuries) [reply frontend/src/Pages/GithubOauth.js:32]: Thanks! Added it in commit [499a073](https://github.com/MatrixDoctors/matrix-cerberus/pull/46/commits/499a073ac58841c19326e4cb6c1844ae4fae23db) under PR #46

File: frontend/src/Components/LayoutsWithNavbar.js (hunk 3 above)
Comment (kuries) [reply frontend/src/Components/LayoutsWithNavbar.js:19]: Done!

https://github.com/MatrixDoctors/matrix-cerberus/blob/398e3af73cfb6f7867b9553e7d9de7d8d8c86300/frontend/src/Components/LayoutsWithNavbar.js#L11-L28

File: frontend/src/Components/LayoutsWithNavbar.js [hunk 5]
```diff
@@ -5,16 +5,18 @@ import { GlobalContext } from '../GlobalContext';
...
 
 export default function LayoutsWithNavbar() {
-  const {setMatrixUserId} = useContext(GlobalContext);
+  const {setMatrixUserId, setGithubUserId} = useContext(GlobalContext);
   const navigate = useNavigate();
 
   useEffect( () => {
       const fetchCurrentUser = async () => {
         const resp = await axios.get('api/current-user');
         console.log(resp.data);
```
[the remark this answers is not in the document]
Comment (kuries) [reply frontend/src/Components/LayoutsWithNavbar.js:14]: I've replaced it with this 

https://github.com/MatrixDoctors/matrix-cerberus/blob/398e3af73cfb6f7867b9553e7d9de7d8d8c86300/frontend/src/Components/LayoutsWithNavbar.js#L14-L19

File: frontend/src/Pages/AccountSettings.js [hunk 6]
```diff
@@ -1,12 +1,22 @@
...
 /**
  * The default web page where an authorized matrix user can edit their third-party account connections.
  */
 export default function AccountSettings() {
-	const {matrixUserId} = useContext(GlobalContext);
+	const {matrixUserId, githubUserId} = useContext(GlobalContext);
 	const homeServer = localStorage.getItem('homeServer');
+
+	function githubHandleClick(){
+		axios.get('api/github/login')
```
[the remark this answers is not in the document]
Comment (kuries) [reply frontend/src/Pages/AccountSettings.js:13]: Thanks for pointing it out!
I've added it in commit [b4484e5](https://github.com/MatrixDoctors/matrix-cerberus/pull/46/commits/b4484e5c2a9e1089baf2f4319f2abea7fc525ad2) under PR #46

File: backend/app/api/deps.py [hunk 7]
```diff
@@ -2,11 +2,37 @@
...
 
 fastapi_sessions = SessionCookie()
 
 
 async def authenticate_user(request: Request):
     if fastapi_sessions.get_session(request) is None:
         raise HTTPException(status_code=401, detail="Not authorized")
+
+
+async def fetch_user_data(session_id, session_data: ServerSessionData):
```
[the remark this answers is not in the document]
Comment (kuries) [reply backend/app/api/deps.py:16]: I've added all the docstrings in this commit [331e407](https://github.com/MatrixDoctors/matrix-cerberus/pull/46/commits/331e407f7a6f49a98748a4fde85658fe6772783d)

File: backend/app/core/parse_events.py [hunk 8]
```diff
@@ -1,8 +1,31 @@
...
+
+    elif type == "rooms":
+        event_type = f"{app_name}.rooms.{additional_type_data['room_id']}"
+
+    elif type == "user":
+        event_type = f"{app_name}.user.{additional_type_data['user_id']}"
+
+    elif type == "global_data":
+        event_type = f"{app_name}.global_data"
+
```
[the remark this answers is not in the document]
Comment (kuries) [reply backend/app/core/parse_events.py:30]: I've raised a ValuError for this case and caught it later with a try-except block.
It's in commit [9017a75](https://github.com/MatrixDoctors/matrix-cerberus/pull/46/commits/9017a75014720ad3366c56b646b9e914f65537c1)

File: backend/app/api/endpoints/github_routes.py [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+
+    params = {"scope": scope, "client_id": client_id, "redirect_uri": redirect_uri, "state": state}
+
+    url = "https://github.com/login/oauth/authorize?"
+    for i in params.keys():
+        url = url + f"{i}={params[i]}&"
+
+    # remove trailing '&'
+    url = url[:-1]
+    print(state)
```
[the remark this answers is not in the document]
Comment (kuries) [reply backend/app/api/endpoints/github_routes.py:33]: Done!

**Pull request merged**